slang.h

```diff
@@ -713,6 +714,32 @@ extern "C"
     };
     #define SLANG_UUID_ISlangFileSystem { 0x003A09FC, 0x3A4D, 0x4BA0, 0xAD, 0x60, 0x1F, 0xD8, 0x63, 0xA9, 0x15, 0xAB }
 
+
+    typedef void(*SlangFuncPtr)(void);
+
+    /** An interface that can be used to encapsulate access to a shared library. An implementaion 
+    does not have to implement the library as a shared library. 
+    */
+    struct ISlangSharedLibrary: public ISlangUnknown
+    {
+        public: 
+            /** Get a function by name. If the library is unloaded will only return nullptr. 
+            @param name The name of the function 
+            @return The function pointer related to the name or nullptr if not found 
+            */
+        virtual SLANG_NO_THROW SlangFuncPtr SLANG_MCALL findFuncByName(char const* name) = 0;
+    };
+    #define SLANG_UUID_ISlangSharedLibrary { 0x9c9d5bc5, 0xeb61, 0x496f,{ 0x80, 0xd7, 0xd1, 0x47, 0xc4, 0xa2, 0x37, 0x30 } };
+
+    struct ISlangSharedLibraryLoader: public ISlangUnknown
+    {
+        public:
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL loadSharedLibrary(
+            const char*     path,
+            ISlangSharedLibrary** sharedLibraryOut) = 0;
```

Comment:
Can you comment on why you chose this design with `ISlangSharedLibrary` over something where `ISlangSharedLibraryLoader` just works with an opaque handle to the loaded library? One consequence of this design is that somebody who wants to implement this interface for themselves needs to deal with allocation of wrapper objects around the `HANDLE` or whatever for shared libraries. We could eliminate the need for the wrapper objects if we eliminated `ISlangSharedLibrary` in favor of `void*` handles.

Reply:
You could just use an opaque handle. But in doing so you have added certain limitations and added a layer of understanding - that has to be documented and perhaps have additional API calls to make it work. 
Specifically - once you know how to use an interface, then whenever you see something that is an interface you know how it works. It may require a little more effort to implement - but what you get back in return is generality and flexibility. This generality/flexibility may not be worth it for light weight objects, or if you had to do it a lot to 'wrap' lots of different types. This is neither.
The generality - is that I can extend and alter my interface/s in well defined ways - through inheritance and through having multiple interfaces. I also know the life time of the thing. There is nothing new or special to add - in terms of managing life times, no special methods to call. I can use pre-existing smart pointer types. In terms of flexibility - I can mix and match implementations. Say I wanted to statically link a library, use regular dll for another, and decompress another. It's trivial. 
You can do so with handles - but you always end up with something that in some sense has to understand all the underlying things, or have things carry around their methods - like a poor mans vtable - with only downside, or enforce that all the things conform to some interface.... but why bother with the handles then?